reinforcement_learning/rlclientlib/utility/watchdog.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+#include <thread>
+#include <map>
+#include <mutex>
+
+#include "api_status.h"
+#include "error_callback_fn.h"
+
+#include <atomic>
+#include "interruptable_sleeper.h"
+
+namespace reinforcement_learning {
+  namespace utility {
+
+
+    class watchdog {
+    public:
+      explicit watchdog(error_callback_fn* error_callback = nullptr);
+      ~watchdog();
+
+      void register_thread(std::thread::id const& thread_id, std::string const& thread_name, long long const timeout);
+      void unregister_thread(std::thread::id const thread_id);
+      void check_in(std::thread::id const thread_id);
+
+      int start(api_status* status);
+      void stop();
+      void loop();
+
+      bool should_report_unhandled_background_error() const;
+      void set_unhandled_background_error(bool const value);
```

Comment:
just bool?

Reply:
Suggest sending the actual actual.

---
In reply to: [213736192](https://github.com/JohnLangford/vowpal_wabbit/pull/1583#discussion_r213736192) [](ancestors = 213736192)

Reply:
I would like to surface the actual error but the problem arises if there are several errors. We would end up shadowing them which I don't think is good behavior. The decision here is to simply notify that there was a background error and to force the user to register a handler and handle them properly

Reply:
OK. As another point then, why have the bool parameter. Is there a scenario where we would want to unset the background error? Maybe flag_unhandled_background_error is more appropriate.

---
In reply to: [215987235](https://github.com/JohnLangford/vowpal_wabbit/pull/1583#discussion_r215987235) [](ancestors = 215987235)